hail/src/main/scala/is/hail/expr/ir/functions/LocusFunctions.scala

```diff
@@ -52,43 +48,45 @@ object LocusFunctions extends RegistryFunctions {
-      val newAlleles = Code.checkcast[IndexedSeq[String]](returnTuple.load().get[java.lang.Object]("_2"))
+        val newLocus = Code.checkcast[Locus](returnTuple.load().get[java.lang.Object]("_1"))
+        val newAlleles = Code.checkcast[IndexedSeq[String]](returnTuple.load().get[java.lang.Object]("_2"))
 
-      val srvb = new StagedRegionValueBuilder(r, PTuple(FastIndexedSeq(tv("T").t.physicalType, PArray(PString()))))
-      Code(
-        returnTuple := tuple,
-        srvb.start(),
-        srvb.addBaseStruct(types.coerce[PBaseStruct](tv("T").t.fundamentalType.physicalType), { locusBuilder =>
-          Code(
-            locusBuilder.start(),
-            locusBuilder.addString(newLocus.invoke[String]("contig")),
-            locusBuilder.advance(),
-            locusBuilder.addInt(newLocus.invoke[Int]("position")))
-        }),
-        srvb.advance(),
-        srvb.addArray(PArray(PString()), { allelesBuilder =>
-          Code(
-            allelesBuilder.start(newAlleles.invoke[Int]("size")),
-            Code.whileLoop(allelesBuilder.arrayIdx < newAlleles.invoke[Int]("size"),
-              allelesBuilder.addString(Code.checkcast[String](newAlleles.invoke[Int, java.lang.Object]("apply", allelesBuilder.arrayIdx))),
-              allelesBuilder.advance()))
-        }),
-        srvb.offset)
+        val newLocusT = PLocus(locusT.rg)
+        val newAllelesT = PArray(PString())
+        val srvb = new StagedRegionValueBuilder(r, PTuple(FastIndexedSeq(newLocusT, newAllelesT)))
+        Code(
+          returnTuple := tuple,
+          srvb.start(),
+          srvb.addBaseStruct(newLocusT.fundamentalType.asInstanceOf[PStruct], { locusBuilder =>
+            Code(
+              locusBuilder.start(),
+              locusBuilder.addString(newLocus.invoke[String]("contig")),
+              locusBuilder.advance(),
+              locusBuilder.addInt(newLocus.invoke[Int]("position")))
+          }),
+          srvb.advance(),
+          srvb.addArray(newAllelesT, { allelesBuilder =>
+            Code(
+              allelesBuilder.start(newAlleles.invoke[Int]("size")),
+              Code.whileLoop(allelesBuilder.arrayIdx < newAlleles.invoke[Int]("size"),
+                allelesBuilder.addString(Code.checkcast[String](newAlleles.invoke[Int, java.lang.Object]("apply", allelesBuilder.arrayIdx))),
+                allelesBuilder.advance()))
+          }),
+          srvb.offset)
     }
 
     registerCode("locus_windows_per_contig", TArray(TArray(TFloat64())), TFloat64(), TTuple(TArray(TInt32()), TArray(TInt32()))) {
-      (r: EmitRegion, coords: Code[Long], radius: Code[Double]) =>
+      case (r: EmitRegion, (groupedT: PArray, coords: Code[Long]), (radiusT: PFloat64, radius: Code[Double])) =>
         val region: Code[Region] = r.region
 
-        val groupedt = PArray(PArray(PFloat64()))
-        val coordt = types.coerce[PArray](groupedt.elementType)
+        val coordT = types.coerce[PArray](groupedT.elementType)
```

Comment:
I also don't quite understand  the original code: 

`val coordt = types.coerce[PArray](groupedt.elementType)`

Isn't the type already PArray? Why do we need to do groupedT.elementType.asInstanceOf[PArray]?

In the changed case 
`val coordT = types.coerce[PArray](groupedT.elementType)`

if groupedT is : PArray[PArray[PFloat64]], then isn't groupedT.elementType guaranteed to by PArray[PFloat64] (in which case no coercion is needed)?

Reply:
The coercion is only a cast so we can access the appropriate code-generation methods on PArray. Otherwise, it's an anonymous PType.

Reply:
Yeah, if it's not possible to specify PArray[PArray[PFloat32]] then the Scala compiler won't understand the specific type of groupedT.elementType